### PR TITLE
types: add export to package.json, test with strict mode and node16 resolution 

### PR DESCRIPTION
**Pull request by ChristianMurphy**
Done:
- [x] use node16 ESM resolution
- [x] add type export to package.json
- [x] test in strict mode
- [x] make sure types match up with implementation (currently failing in `strict` mode)

resolves https://github.com/Rich-Harris/estree-walker/issues/35
resolves https://github.com/Rich-Harris/estree-walker/issues/28
resolves https://github.com/Rich-Harris/estree-walker/pull/24

### CODE REVIEW

File: src/async.js [hunk 1]
```diff
@@ -1,40 +1,56 @@
...
+
+		/** @type {Node | null} */
+		this.replacement = null;
+
+		/** @type {WalkerContext} */
+		this.context = {
+			skip: () => (this.should_skip = true),
+			remove: () => (this.should_remove = true),
+			replace: (node) => (this.replacement = node)
+		};
```
Comment (ChristianMurphy) [inline src/async.js:38]: These shouldn't need to be here, they are defined by the `super()` calling `WalkerBase`.
But TypeScript + JSDoc seems to be unable to infer these from the `extends WalkerBase` on the class. 🤔 

/cc @wooorm @remcohaszing have you ever run into this before? Do you know of any work arounds?

Comment (remcohaszing) [reply src/async.js:38]: Apparently the culprit are these lines in `visit`:

```js
const _should_skip = this.should_skip;
const _should_remove = this.should_remove;
const _replacement = this.replacement;

// … 

this.should_skip = _should_skip;
this.should_remove = _should_remove;
this.replacement = _replacement;
```

It appears TypeScript is having problems recursively inferring these types from themselves. As a workaround, you can type-cast the values that are later assigned back. Note that this occurs twice!

```js
const _should_skip = /** @type {boolean} */ (this.should_skip);
const _should_remove = /** @type {boolean} */ (this.should_remove);
const _replacement = /** @type {Node | null} */ (this.replacement);
```

---

This is definitely a bug in TypeScript. Here’s minimal reproduction. See what happens if you comment the line `this.name = name`.

```js
class Foo {
  name = ''
}

class Bar extends Foo {
  rename() {
    const name = this.name
    this.name = name
  }
}
```

File: src/async.js [hunk 2]
```diff
@@ -1,40 +1,56 @@
...
+
+		/** @type {AsyncHandler | undefined} */
```
Comment (remcohaszing) [inline src/async.js:40]: If you assign `this.whatever = whatever` in the constructor, TypeScript can infer its type, meaning this type annotation is redundant. There are multiple occurrences across multiple files. It’s not wrong, but personally I would omit them.

File: tsconfig.json [hunk 3]
```diff
@@ -1,6 +1,9 @@
 {
 	"compilerOptions": {
 		"allowJs": true,
+		"checkJs": true,
```
Comment (remcohaszing) [inline tsconfig.json:4]: `checkJs` implies `allowJs`

```suggestion
		"checkJs": true,
```

File: src/walker.js [hunk 4]
```diff
@@ -27,32 +27,32 @@ export class WalkerBase {
...
+	 * @param {Parent} parent
+	 * @param {keyof Parent} prop
+	 * @param {number | null} index
+	 * @param {Node} node
 	 */
 	replace(parent, prop, index, node) {
 		if (parent) {
-			if (index !== null) {
-				parent[prop][index] = node;
+			if (index !== null && typeof index !== 'undefined') {
```
Comment (remcohaszing) [inline src/walker.js:38]: Personally I prefer the following equivalent:

```suggestion
			if (index != null) {
```

File: package.json [hunk 5]
```diff
@@ -25,7 +25,7 @@
 	},
 	"devDependencies": {
 		"@types/estree": "0.0.42",
```
Comment (remcohaszing) [inline package.json:27]: `@types/estree` is needed for the type definitions as part of the public interface, so it belongs in `dependencies`, not `devDependencies`. Also I suggest to just update this to `^1.0.0` while you’re at it.

File: src/async.js (hunk 1 above)
Comment (wooorm) [reply src/async.js:38]: Maybe you need to type fields? Such as like this: https://github.com/vfile/vfile/blob/03efac7cc3a67fe33bb7a0f1de29e17be322549e/lib/index.js#L122-L126

Comment (remcohaszing) [reply src/async.js:38]: No, it’s currently done that way, but it doesn’t make a difference. It’s a bug with subclasses. Also it only happens in JS. If the exact same code in the minimal reproduction is added in a `.ts` file, TypeScript can infer the types perfectly fine.

**Comment by wooorm** [comment]
@ChristianMurphy What’s blocking this?

**Comment by ChristianMurphy** [comment]
> What’s blocking this?

Time, more specifically a lack thereof.
This is mostly ready, that last remaining item here is unraveling some type inconsistencies in the `visit` method in _sync.js_ and _async.js_ which are a bit complex.

I think there are two separate things which need to be resolved in the `visit` method.
1. Resolve inconsistency on whether params are optional or not, externally they appear to be optional, internally they appear to be required.
2. Sort out some edge cases in the `for (const key in node) {` loop, where source location and comments potentially break assumptions in the code.

After those are resolve it would be nice to get https://github.com/Rich-Harris/estree-walker/pull/34#discussion_r1062455750 sorted, which will likely require an issue upstream at TypeScript, some back and forth, and some lead time waiting for a release.

I'll aim to keep chipping away at the remaining items.
If you or others have time to pitch in, contributions are welcome!
And would help speed the timeline up.

File: package.json (hunk 5 above)
Comment (ChristianMurphy) [reply package.json]: Good idea, updated

Comment (ChristianMurphy) [reply package.json]: this also resolves https://github.com/Rich-Harris/estree-walker/issues/35

**Comment by wooorm** [comment]
Here’s a diff to make the types and tests work:

```diff
diff --git a/src/async.js b/src/async.js
index 64f6933..f068c71 100644
--- a/src/async.js
+++ b/src/async.js
@@ -6,9 +6,9 @@ import { WalkerBase } from './walker.js';
  * @typedef {(
  *    this: WalkerContext,
  *    node: Node,
- *    parent: Node,
- *    key: string | number | symbol,
- *    index: number
+ *    parent: Node | null,
+ *    key: string | number | symbol | null | undefined,
+ *    index: number | null | undefined
  * ) => Promise<void>} AsyncHandler
  */
 
@@ -47,9 +47,9 @@ export class AsyncWalker extends WalkerBase {
 	/**
 	 * @template {Node} Parent
 	 * @param {Node} node
-	 * @param {Parent} parent
-	 * @param {keyof Parent} prop
-	 * @param {number} index
+	 * @param {Parent | null} parent
+	 * @param {keyof Parent} [prop]
+	 * @param {number | null} [index]
 	 * @returns {Promise<Node | null>}
 	 */
 	async visit(node, parent, prop, index) {
@@ -84,22 +84,28 @@ export class AsyncWalker extends WalkerBase {
 				if (removed) return null;
 			}
 
-			for (const key in node) {
-				const value = node[/** @type {keyof Node} */ (key)];
-
-				if (typeof value !== "object") {
-					continue;
-				} else if (Array.isArray(value)) {
-					for (let i = 0; i < value.length; i += 1) {
-						if (value[i] !== null && typeof value[i].type === 'string') {
-							if (!(await this.visit(value[i], node, key, i))) {
-								// removed
-								i--;
+			/** @type {keyof Node} */
+			let key;
+
+			for (key in node) {
+				/** @type {unknown} */
+				const value = node[key];
+
+				if (value && typeof value === 'object') {
+					if (Array.isArray(value)) {
+						const nodes = /** @type {Array<unknown>} */ (value);
+						for (let i = 0; i < nodes.length; i += 1) {
+							const item = nodes[i];
+							if (isNode(item)) {
+								if (!(await this.visit(item, node, key, i))) {
+									// removed
+									i--;
+								}
 							}
 						}
+					} else if (isNode(value)) {
+						await this.visit(value, node, key, null);
 					}
-				} else if (value !== null && typeof value.type === "string") {
-					await this.visit(value, node, key, null);
 				}
 			}
 
@@ -132,3 +138,15 @@ export class AsyncWalker extends WalkerBase {
 		return node;
 	}
 }
+
+/**
+ * Ducktype a node.
+ *
+ * @param {unknown} value
+ * @returns {value is Node}
+ */
+function isNode(value) {
+	return (
+		value !== null && typeof value === 'object' && 'type' in value && typeof value.type === 'string'
+	);
+}
diff --git a/src/sync.js b/src/sync.js
index a7ad898..171fb36 100644
--- a/src/sync.js
+++ b/src/sync.js
@@ -6,9 +6,9 @@ import { WalkerBase } from './walker.js';
  * @typedef {(
  *    this: WalkerContext,
  *    node: Node,
- *    parent: Node,
- *    key?: string | number | symbol | undefined,
- *    index?: number | undefined
+ *    parent: Node | null,
+ *    key: string | number | symbol | null | undefined,
+ *    index: number | null | undefined
  * ) => void} SyncHandler
  */
 
@@ -47,9 +47,9 @@ export class SyncWalker extends WalkerBase {
 	/**
 	 * @template {Node} Parent
 	 * @param {Node} node
-	 * @param {Parent} parent
-	 * @param {keyof Parent} prop
-	 * @param {number} index
+	 * @param {Parent | null} parent
+	 * @param {keyof Parent} [prop]
+	 * @param {number | null} [index]
 	 * @returns {Node | null}
 	 */
 	visit(node, parent, prop, index) {
@@ -84,22 +84,28 @@ export class SyncWalker extends WalkerBase {
 				if (removed) return null;
 			}
 
-			for (const key in node) {
-				const value = node[/** @type {keyof Node} */(key)];
-
-				if (typeof value !== "object") {
-					continue;
-				} else if (Array.isArray(value)) {
-					for (let i = 0; i < value.length; i += 1) {
-						if (value[i] !== null && typeof value[i].type === 'string') {
-							if (!this.visit(value[i], node, key, i)) {
-								// removed
-								i--;
+			/** @type {keyof Node} */
+			let key;
+
+			for (key in node) {
+				/** @type {unknown} */
+				const value = node[key];
+
+				if (value && typeof value === 'object') {
+					if (Array.isArray(value)) {
+						const nodes = /** @type {Array<unknown>} */ (value);
+						for (let i = 0; i < nodes.length; i += 1) {
+							const item = nodes[i];
+							if (isNode(item)) {
+								if (!this.visit(item, node, key, i)) {
+									// removed
+									i--;
+								}
 							}
 						}
+					} else if (isNode(value)) {
+						this.visit(value, node, key, null);
 					}
-				} else if (value !== null && typeof value.type === "string") {
-					this.visit(value, node, key, null);
 				}
 			}
 
@@ -132,3 +138,15 @@ export class SyncWalker extends WalkerBase {
 		return node;
 	}
 }
+
+/**
+ * Ducktype a node.
+ *
+ * @param {unknown} value
+ * @returns {value is Node}
+ */
+function isNode(value) {
+	return (
+		value !== null && typeof value === 'object' && 'type' in value && typeof value.type === 'string'
+	);
+}
diff --git a/src/walker.js b/src/walker.js
index cfbc406..0c54bb9 100644
--- a/src/walker.js
+++ b/src/walker.js
@@ -28,31 +28,31 @@ export class WalkerBase {
 
 	/**
 	 * @template {Node} Parent
-	 * @param {Parent} parent
-	 * @param {keyof Parent} prop
-	 * @param {number | null} index
+	 * @param {Parent | null | undefined} parent
+	 * @param {keyof Parent | null | undefined} prop
+	 * @param {number | null | undefined} index
 	 * @param {Node} node
 	 */
 	replace(parent, prop, index, node) {
-		if (parent) {
+		if (parent && prop) {
 			if (index !== null && typeof index !== 'undefined') {
-				/** @type {Array<Node>} */(parent[prop])[index] = node;
+				/** @type {Array<Node>} */ (parent[prop])[index] = node;
 			} else {
-				/** @type {Node} */(parent[prop]) = node;
+				/** @type {Node} */ (parent[prop]) = node;
 			}
 		}
 	}
 
 	/**
 	 * @template {Node} Parent
-	 * @param {Parent} parent
-	 * @param {keyof Parent} prop
-	 * @param {number} index
+	 * @param {Parent | null | undefined} parent
+	 * @param {keyof Parent | null | undefined} prop
+	 * @param {number | null | undefined} index
 	 */
 	remove(parent, prop, index) {
-		if (parent) {
-			if (index !== null) {
-				/** @type {Array<Node>} */(parent[prop]).splice(index, 1);
+		if (parent && prop) {
+			if (index !== null && index !== undefined) {
+				/** @type {Array<Node>} */ (parent[prop]).splice(index, 1);
 			} else {
 				delete parent[prop];
 			}
```

**Comment by wooorm** [comment]
> where source location and comments potentially break assumptions in the code.

This is an intentional decision I believe here: to allow arbitrary and even unknown nodes instead of tracking which nodes are valid. Otherwise you’d need a big list like https://github.com/eslint/eslint-visitor-keys/tree/main/lib.

**Comment by ChristianMurphy** [comment]
That did the trick, thanks @wooorm! Applied in https://github.com/Rich-Harris/estree-walker/pull/34/commits/52e383977bc06e0cc81d761a0c2fe0b27acca7e2

File: src/walker.js (hunk 4 above)
Comment (ChristianMurphy) [reply src/walker.js]: Sure, applied

**Comment by ChristianMurphy** [comment]
@Rich-Harris this is ready for a review when you have a moment

**Review by wooorm — APPROVED** [review]

File: src/sync.js [hunk 6]
```diff
@@ -116,3 +138,15 @@ export class SyncWalker extends WalkerBase {
...
+ * Ducktype a node.
+ *
+ * @param {unknown} value
+ * @returns {value is Node}
+ */
+function isNode(value) {
+	return (
+		value !== null && typeof value === 'object' && 'type' in value && typeof value.type === 'string'
+	);
+}
```
Comment (remcohaszing) [inline src/sync.js:152]: This function is defined twice. (It’s also in `async.js`.) I suggest to deduplicate this logic.

Comment (wooorm) [reply src/sync.js:152]: a lot of the code is duplicated tho

**Review by Rich-Harris — APPROVED** [review]

**Comment by Rich-Harris** [comment]
released 3.0.3. thank you!

**Comment by ChristianMurphy** [comment]
Thanks @Rich-Harris! 🙇

**Pull request merged**